tests/integration/test_charm.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from pathlib import Path
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+from tests.integration.helpers import pull_content_from_unit_file
+from pymongo import MongoClient
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy one unit of MongoDB"""
+    my_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(my_charm)
+    await ops_test.model.wait_for_idle()
+
+
+@pytest.mark.abort_on_fail
+async def test_status(ops_test: OpsTest) -> None:
+    """Verifies that the application and unit are active"""
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", timeout=1000
+    )
+    assert len(ops_test.model.applications[APP_NAME].units) == 1
+    assert ops_test.model.applications[APP_NAME].units[0].workload_status == "active"
+
+
+@pytest.mark.abort_on_fail
+async def test_config_files_are_correct(ops_test: OpsTest) -> None:
+    """Tests that mongo.conf as expected content"""
+    # Get the expected contents from files.
+    with open("tests/data/mongod.conf") as file:
+        expected_mongodb_conf = file.read()
+
+    # Pull the configuration files from each PostgreSQL instance.
+    for unit in ops_test.model.applications[f"{APP_NAME}"].units:
+        # Check that the remaining settings are as expected.
+        unit_mongodb_conf_data = await pull_content_from_unit_file(
+            unit, "/etc/mongod.conf"
+        )
+        expected_mongodb_conf = update_bind_ip(
+            expected_mongodb_conf, unit.public_address
+        )
+        assert expected_mongodb_conf == unit_mongodb_conf_data
+
+
+@pytest.mark.abort_on_fail
+async def test_database_is_up_as_replica_set(ops_test: OpsTest) -> None:
+    """Tests that mongodb is running as a replica set for the application units"""
+    # construct URI for MongoDB replica set
+    uri = "mongodb://"
+    for i in range(0, len(ops_test.model.applications[APP_NAME].units)):
+        if i:
+            uri += ","
+        unit = ops_test.model.applications[APP_NAME].units[i]
+        uri += "{}:{}".format(unit.public_address, 27017)
+    uri += "/?replicaSet=rs0"
+
+    # connect to mongo replicaSet
+    client = MongoClient(uri)
+
+    # check mongo replicaset is ready
+    assert client.server_info()
```

Comment:
Maybe you can check the return against an expected content here too, to make it clear what is being checked (I believe it's a string).

Reply:
Yes! Actually I think that `client.server_info()` returns a string when the server is ready and otherwise it throws a `ServerSelectionTimeoutError` do you think I should just check that it doesn't throw a `ServerSelectionTimeoutError`?

Reply:
This might help: https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions

Reply:
I was thinking about checking the content of the string (just to confirm that it returns the expected content based on the url that was passed to the client). But, an additional check like you and Jon mentioned would be very nice (for testing with an invalid url).

Reply:
@marceloneppel I was thinking of doing something like this at first, but then I realised that the information returned by `server_info` gives information about the server that MongoDB is running on, such as the version of MongoDB, the version of git, what javascriptEngine is running, what the build environment is, etc. It doesn't really provide anything that is related to the deployment of MongoDB. 

**But** by returning this information and not throwing an error, it verifies that we were able to make a connection with MongoDB on both the units specified **and** the replicaset specified. This verifies two things:
1. an instance of MongoDB is running on the specified unit
2. the replicaset specified does exist **and** we can reach it

If we wanted to do further checks we could ask the client for information with something like `client.admin.command("replSetGetConfig")` in which we pass the command `replSetGetConfig` to the MongoDB client which we hopefully give us the config options for the replica set. And we could check that information to verify that specified replica set is there and that it runs for each unit. But I think this would be redundant since the fact we get a response from the server indicates these things already. 

I could go either way on this  🤷 , let me know if you want to do these extra checks. I trust your judgement 😄

Reply:
Hi Mia! Yes. I agree that checking that information that you described is redundant. With all this info now, I believe we can keep the test as they are. Thanks for the details, Mia!

Reply:
@jnsgruk Thanks for this link! It's quite helpful about asserting that an exception did occur. Since I am in the case of trying to assert that an exception did not occur. It makes things more complicated since we can't use the `with pytest.raises(ServerSelectionTimeoutError):` I think it might be the case that we would have to try and if an exception is thrown then we assert False with an explaination. Or am I missing something. Sorry if it is obvious I'm still experiencing some fatigue and brainfog